### ISSUE TITLE
setgroups() raises when given too many groups, and nothing handles the error

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10412]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#600|
|Type|defect|
|Created|2004-04-18 02:14:43Z|

Attachments:

* [setgroups.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/15a/15aa0c7e8fbd2923db7041d012e8838d66b9572d/eebc0729b1cceb8ff1628b2cc5df41b34cbe4458.patch) (2520 bytes) - added by exarkun on 2004-12-21 11:06:39Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__600 600
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1082254483000000 1082254483000000
changetime__1480617823933097 1480617823933097
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__itamarst cc__Tv cc__moshez cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
2004/04/17 14:11 MDT [-]   File
"/home/radix/Projects/Twisted/twisted/scripts/twistd.py", line 160, in
startApplication
2004/04/17 14:11 MDT [-]     shedPrivileges(config['euid'], process.uid,
process.gid)
2004/04/17 14:11 MDT [-]   File
"/home/radix/Projects/Twisted/twisted/scripts/twistd.py", line 134, in
shedPrivileges
2004/04/17 14:11 MDT [-]     switchUID(uid, gid, euid)
2004/04/17 14:11 MDT [-]   File
"/home/radix/Projects/Twisted/twisted/python/util.py", line 600, in switchUID
2004/04/17 14:11 MDT [-]     initgroups(uid, gid)
2004/04/17 14:11 MDT [-]   File
"/home/radix/Projects/Twisted/twisted/python/util.py", line 578, in initgroups
2004/04/17 14:11 MDT [-]     setgroups(l)
2004/04/17 14:11 MDT [-] ValueError: too many groups
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Is this bug still outstanding?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Tv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Tv commented|
|-|-|

```
#!html
<pre>
Yes and no. As much as it ever was :)

From the bug report:

"/home/radix/Projects/Twisted/twisted/python/util.py", line 578, in initgroups
2004/04/17 14:11 MDT [-]     setgroups(l)
2004/04/17 14:11 MDT [-] ValueError: too many groups

There is no code in twisted.python.util that would protect that
setgroups from raising ValueErrors to the caller.

Then again, this is a case of running against an OS-imposed limit,
and I think that code MUST fail. Whether it should catch ValueError
and raise something like TooManyGroupsError is another issue.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Here's part of a "fix".

I think the rest of the fix comes in two parts:

  Don't even try to setgroups() if uid != 0

  If Application is passed None for uid and gid, don't even call initgroups().
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Target the uid/gid changes for 2.0.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Resolved by 13136
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Tv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Tv commented|
|-|-|

```
#!html
<pre>
How is this fixed by that r13136? I don't see it doing anything else except not
calling switchUID by default, and thus, making the bug appear less often.
If you use --uid or --gid, and the users belongs to too many groups, it'll still
fail in exactly the same way.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
OMG, this is the code from glibc in initgroups():
  do
    result = setgroups (ngroups, groups);
  while (result == -1 &amp;&amp; errno == EINVAL &amp;&amp; --ngroups > 0);

That is fucking evil. I'll commit a fix to twisted to do the same.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Tv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Tv commented|
|-|-|

```
#!html
<pre>
Recording this just for future reference: the commit that actually "fixed"
(read: covered up) things was r13140. Sorry for the noise, but I want that here
or I'll keep re-searching the right commit all the time.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_11"></a>|@rodrigc commented|
|-|-|

In changeset 91ff8029666decf9367bc72de12e21c53dc26736
```
#!CommitTicketReference repository="" revision="91ff8029666decf9367bc72de12e21c53dc26736"
Merge pull request #600 from twisted/8913-htmlizer-rodrigc

Author: rodrigc
Reviewer: glyph
Fixes: #8913

Port pyhtmlizer to Python 3
```
